OpenSim/Simulation/Test/testInverseKinematicsSolver.cpp

```diff
@@ -0,0 +1,457 @@
+ * -------------------------------------------------------------------------- */
+
+//=============================================================================
+// testInverseKinematicsSolver verifies that changes to the accuracy and marker
+// weights have expected effects on the inverse kinematics results/errors
+//=============================================================================
+#include <OpenSim/Simulation/osimSimulation.h>
+#include <OpenSim/Simulation/InverseKinematicsSolver.h>
+#include <OpenSim/Simulation/MarkersReference.h>
+#include <OpenSim/Common/Constant.h>
+#include <OpenSim/Common/STOFileAdapter.h>
+#include <random>
+
+using namespace OpenSim;
+using namespace std;
+
+// Utility function to build a simple pendulum with markers attached
+Model* constructPendulumWithMarkers();
+// Using a model with markers and trajectory of states, create synthetic
+// marker data. If noiseRadius is provided use it to scale the noise
+// that perturbs the marker data. Optional fixed parameter to use the same
+// noise for each time frame or to compute noise to be added at each frame.
+MarkerData* generateMarkerDataFromModelAndStates(const Model& model,
+    const StatesTrajectory& states, double noiseRadius=0, bool fixed=false);
+
+// Verify that accuracy improves the number of decimals points to which
+// the solver solution (coordinates) can be trusted as it is tightened.
+void testAccuracy();
+// Verify that the marker weights impact the solver and has the expected
+// effect of reducing the error for the marker weight that is increased. 
+void testUpdateMarkerWeights();
+// Verify that the track() solution is also effected by updating marker
+// weights and marker error is being reduced as its weighting increases.
+void testTrackWithUpdateMarkerWeights();
+
+int main()
+{
+    SimTK::Array_<std::string> failures;
+
+    try { testAccuracy(); }
+    catch (const std::exception& e) {
+        cout << e.what() << endl; failures.push_back("testAccuracy");
+    }
+    try { testUpdateMarkerWeights(); }
+    catch (const std::exception& e) {
+        cout << e.what() << endl; failures.push_back("testUpdateGoalWeights");
+    }
+    try { testTrackWithUpdateMarkerWeights(); }
+    catch (const std::exception& e) {
+        cout << e.what() << endl; failures.push_back("testAccuracy");
+    }
+
+    if (!failures.empty()) {
+        cout << "Done, with failure(s): " << failures << endl;
+        return 1;
+    }
+
+    cout << "Done. All cases passed." << endl;
+
+    return 0;
+}
+
+//=============================================================================
+// Test Cases
+//=============================================================================
+
+void testAccuracy()
+{
+    cout << "\ntestInverseKinematicsSolver::testAccuracy()" << endl;
+
+    std::unique_ptr<Model> pendulum{ constructPendulumWithMarkers() };
+    Coordinate& coord = pendulum->getCoordinateSet()[0];
+
+    double refVal = 0.123456789;
```

Comment:
`refVal0` wasn't necessary, so it has been removed. I used `refVal` for its correspondence with it being the `CoordinateReference` value.  `coordVal` was intended to indicate the model's coordinate value. Would you prefer `referenceValue` and `modelValue` (since they are both Coordinate values)?